### ISSUE TITLE
tf 2.2.0 achieves much worse results than tf 2.1.0, same model

### ISSUE THREAD

**Issue by notnot** [tensorflow/tensorflow#41513]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1.  It must be a bug, a feature request, or a significant problem with the
    documentation (for small docs fixes please send a PR instead).
2.  The form below must be filled out.
3.  It shouldn't be a TensorBoard issue. Those go
    [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**:
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
    happens on a mobile device**:
-   **TensorFlow installed from (source or binary)**:
-   **TensorFlow version (use command below)**:
-   **Python version**:
-   **Bazel version (if compiling from source)**:
-   **GCC/Compiler version (if compiling from source)**:
-   **CUDA/cuDNN version**:
-   **GPU model and memory**:
-   **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:

```bash
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"
```

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by notnot** [comment]
Have I written custom code (as opposed to using a stock example script
provided in TensorFlow):

Yes, i have custom code.

OS Platform and Distribution (e.g., Linux Ubuntu 16.04):

Ubunto 18.04

Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
happens on a mobile device:

NA

TensorFlow installed from (source or binary):

pip3 install tensorflow==2.2

TensorFlow version (use command below):

2.2 (versus 2.1)

Python version:

3.6.9


Bazel version (if compiling from source):

NA

GCC/Compiler version (if compiling from source):

NA

CUDA/cuDNN version:

CUDA version 11.0 according to nvidia-smi

GPU model and memory:

Jetson Nano (Tegra X1)

Exact command to reproduce:

NA (running my custom tensorflow application)




So... i have a transformer model which was achieving great results with TensorFlow 2.1, and after updating to TensorFlow 2.2, the results are bad.

**Comment by lordtt13** [comment]
Does the 2.2 version recognize GPU?

**Comment by lordtt13** [comment]
By bad results do you mean training takes longer or inferences are bad?

**Comment by notnot** [comment]
Yes GPU is recognized. Inferences are bad :(

**Comment by lordtt13** [comment]
Can you link a notebook showing infrences in both versions?

**Comment by notnot** [comment]
I could make it a single-file script and include it right here?
No idea how to make a notebook that can use a specific version of tensorflow.

**Comment by lordtt13** [comment]
Yes please do share, I'll run it with both versions

**Comment by notnot** [comment]
Ok, here is a repro case. This is a new project, a work in progress. The aim of the network is to predict the next word,
given a sequence of words. After a very short training the script will display pairs of input and output sequences. When using tf 2.1, the results are usually perfect. Not so with tf 2.2, as you will see.

The source code of the script is between the horizontal lines.

--------------------------------------------------------------------------------

# trans.py, jpad 2020
# transformer network, based on: 
# https://machinetalk.org/2019/04/29/create-the-transformer-with-tensorflow-2-0/

import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'  # level in [0, 3]

import numpy as np
np.set_printoptions(suppress=True) # avoid scientific notation

import time
import unicodedata
import re
import sys

import tensorflow as tf
print("TensorFlow version", tf.__version__)


#### data preprocessing ########################################################

SEQ_LEN = 12

text_data = '''A long time ago there were a king and queen who said every day, 
"Ah, if only we had a child," but they never had one.
'''

def is_punctuation(c):
    cat = unicodedata.category(c)
    return cat=='Po' or cat=='Pc' or cat=='Pi' or cat=='Pf' or cat=='Ps' or cat=='Pe'

def is_letter(c):
    cat = unicodedata.category(c)
    return cat=='Lu' or cat=='Ll' or cat=='Lt' or cat=='Lm' or cat=='Lo'

def is_number(c):
    cat = unicodedata.category(c)
    return cat=='Nd' or cat=='Nl' or cat=='No'

def separate_words(s):
    '''
    Add whitespace around punctuation marks. Returns the string as a list of
    words.
    '''
    chars = [c for c in unicodedata.normalize('NFD', s)]
    s = ''
    for i, c in enumerate(chars):
        if is_punctuation(c):
            if i > 0:
                # if previous character is alphanumerical: add space
                pc = chars[i-1]
                if is_letter(pc) or is_number(pc):
                    s += " "

            s += c
            
            if i < len(chars) -1:
                # if next character is not a separator, add a space
                nc = chars[i+1]
                if nc != " ":
                    s += " "
        else:
            s += c
        
    return s.split()


def make_old_new(source, seq_len):
    # remove newlines
    source = source.replace('\n', '')
    # split into 'words' between spaces and punctuation marks
    words = separate_words(source)

    olds = []
    news = []
    n_words = len(words)
    # make fragments
    for i in range(n_words - seq_len + 1):
        fragment = words[i:i+seq_len]
        string = ' '.join(fragment)
        olds.append(string)

    # new from old
    for i in range(len(olds)-1):
        news.append(olds[i+1])
    olds = olds[:-1] # last old fragment can't be coupled to a new one

    return olds, news


# text preprocessing
raw_data_old, raw_data_new = make_old_new(text_data, SEQ_LEN)
print("old:\n", raw_data_old)
print("new:\n", raw_data_new)

raw_data_new_in = ['<start> ' + data for data in raw_data_new]
raw_data_new_out = [data + ' <end>' for data in raw_data_new]

# tokenise
old_tokenizer = tf.keras.preprocessing.text.Tokenizer(filters='')
old_tokenizer.fit_on_texts(raw_data_old)
print("word index (old):\n", old_tokenizer.word_index)
data_old = old_tokenizer.texts_to_sequences(raw_data_old)
data_old = tf.keras.preprocessing.sequence.pad_sequences(
    data_old, padding='post')

new_tokenizer = tf.keras.preprocessing.text.Tokenizer(filters='')
new_tokenizer.fit_on_texts(raw_data_new_in)
new_tokenizer.fit_on_texts(raw_data_new_out)
data_new_in = new_tokenizer.texts_to_sequences(raw_data_new_in)
data_new_in = tf.keras.preprocessing.sequence.pad_sequences(
    data_new_in, padding='post')
data_new_out = new_tokenizer.texts_to_sequences(raw_data_new_out)
data_new_out = tf.keras.preprocessing.sequence.pad_sequences(
    data_new_out, padding='post')

# create a standard tensorflow dataset
BATCH_SIZE = 4
dataset = tf.data.Dataset.from_tensor_slices(
    (data_old, data_new_in, data_new_out))
dataset = dataset.shuffle(len(raw_data_old)).batch(BATCH_SIZE)


#### positional encoding #######################################################

MODEL_SIZE = 128

def positional_encoding(pos, model_size):
    """ Compute positional encoding for a particular position
    Args:
        pos: position of a token in the sequence
        model_size: depth size of the model
    
    Returns:
        The positional encoding for the given token
    """
    PE = np.zeros((1, model_size))
    for i in range(model_size):
        if i % 2 == 0:
            PE[:, i] = np.sin(pos / 10000 ** (i / model_size))
        else:
            PE[:, i] = np.cos(pos / 10000 ** ((i - 1) / model_size))
    return PE


pes = []
max_length = max(len(data_old[0]), len(data_new_in[0]))
for i in range(max_length):
    pes.append(positional_encoding(i, MODEL_SIZE))

pes = np.concatenate(pes, axis=0)
pes = tf.constant(pes, dtype=tf.float32)

print("pes shape", pes.shape)


#### attention #################################################################

class MultiHeadAttention(tf.keras.Model):
    """ Class for Multi-Head Attention layer
    Attributes:
        key_size: d_key in the paper
        h: number of attention heads
        wq: the Linear layer for Q
        wk: the Linear layer for K
        wv: the Linear layer for V
        wo: the Linear layer for the output
    """
    def __init__(self, model_size, h):
        super(MultiHeadAttention, self).__init__()
        self.key_size = model_size // h
        self.h = h
        self.wq = tf.keras.layers.Dense(model_size) 
        self.wk = tf.keras.layers.Dense(model_size) 
        self.wv = tf.keras.layers.Dense(model_size) 
        self.wo = tf.keras.layers.Dense(model_size)

    def call(self, query, value, mask=None):
        """ The forward pass for Multi-Head Attention layer
        Args:
            query: the Q matrix
            value: the V matrix, acts as V and K
            mask: mask to filter out unwanted tokens
                - zero mask: mask for padded tokens
                - right-side mask: mask to prevent attention towards tokens 
                  on the right-hand side
        
        Returns:
            The concatenated context vector
            The alignment (attention) vectors of all heads
        """
        # query has shape (batch, query_len, model_size)
        # value has shape (batch, value_len, model_size)
        query = self.wq(query)
        key = self.wk(value)
        value = self.wv(value)
        
        # Split matrices for multi-heads attention
        batch_size = query.shape[0]
        
        # Originally, query has shape (batch, query_len, model_size)
        # We need to reshape to (batch, query_len, h, key_size)
        query = tf.reshape(query, [batch_size, -1, self.h, self.key_size])
        # In order to compute matmul, the dimensions must be transposed to (batch, h, query_len, key_size)
        query = tf.transpose(query, [0, 2, 1, 3])
        
        # Do the same for key and value
        key = tf.reshape(key, [batch_size, -1, self.h, self.key_size])
        key = tf.transpose(key, [0, 2, 1, 3])
        value = tf.reshape(value, [batch_size, -1, self.h, self.key_size])
        value = tf.transpose(value, [0, 2, 1, 3])
        
        # Compute the dot score
        # and divide the score by square root of key_size (as stated in paper)
        # (must convert key_size to float32 otherwise an error would occur)
        score = tf.matmul(query, key, transpose_b=True) / tf.math.sqrt(tf.dtypes.cast(self.key_size, dtype=tf.float32))
        # score will have shape of (batch, h, query_len, value_len)
        
        # Mask out the score if a mask is provided
        # There are two types of mask:
        # - Padding mask (batch, 1, 1, value_len): to prevent attention being drawn to padded token (i.e. 0)
        # - Look-left mask (batch, 1, query_len, value_len): to prevent decoder to draw attention to tokens to the right
        if mask is not None:
            score *= mask

            # We want the masked out values to be zeros when applying softmax
            # One way to accomplish that is assign them to a very large negative value
            score = tf.where(tf.equal(score, 0), tf.ones_like(score) * -1e9, score)
        
        # Alignment vector: (batch, h, query_len, value_len)
        alignment = tf.nn.softmax(score, axis=-1)
        
        # Context vector: (batch, h, query_len, key_size)
        context = tf.matmul(alignment, value)
        
        # Finally, do the opposite to have a tensor of shape (batch, query_len, model_size)
        context = tf.transpose(context, [0, 2, 1, 3])
        context = tf.reshape(context, [batch_size, -1, self.key_size * self.h])
        
        # Apply one last full connected layer (WO)
        heads = self.wo(context)
        
        return heads, alignment


#### encoder ###################################################################

class Encoder(tf.keras.Model):
    """ Class for the Encoder
    Args:
        vocab_size: vocabulary size of the input
        model_size: d_model in the paper (depth size of the model)
        num_layers: number of layers (Multi-Head Attention + FNN)
        h: number of attention heads
        pes: positional emcodings

        embedding: Embedding layer
        embedding_dropout: Dropout layer for Embedding
        attention: array of Multi-Head Attention layers
        attention_dropout: array of Dropout layers for Multi-Head Attention
        attention_norm: array of LayerNorm layers for Multi-Head Attention
        dense_1: array of first Dense layers for FFN
        dense_2: array of second Dense layers for FFN
        ffn_dropout: array of Dropout layers for FFN
        ffn_norm: array of LayerNorm layers for FFN
    """
    def __init__(self, vocab_size, model_size, num_layers, h, pes):
        super(Encoder, self).__init__()
        self.model_size = model_size
        self.num_layers = num_layers
        self.h = h
        self.pes = pes
        self.embedding = tf.keras.layers.Embedding(vocab_size, model_size)
        self.embedding_dropout = tf.keras.layers.Dropout(0.1)
        self.attention = [MultiHeadAttention(model_size, h) for _ in range(num_layers)]
        self.attention_dropout = [tf.keras.layers.Dropout(0.1) for _ in range(num_layers)]

        self.attention_norm = [tf.keras.layers.LayerNormalization(
            epsilon=1e-6) for _ in range(num_layers)]

        self.dense_1 = [tf.keras.layers.Dense(
            model_size * 4, activation='relu') for _ in range(num_layers)]
        self.dense_2 = [tf.keras.layers.Dense(
            model_size) for _ in range(num_layers)]
        self.ffn_dropout = [tf.keras.layers.Dropout(0.1) for _ in range(num_layers)]
        self.ffn_norm = [tf.keras.layers.LayerNormalization(
            epsilon=1e-6) for _ in range(num_layers)]

    def call(self, sequence, training=True, encoder_mask=None):
        """ Forward pass for the Encoder
        Args:
            sequence: source input sequences
            training: whether training or not (for Dropout)
            encoder_mask: padding mask for the Encoder's Multi-Head Attention
        
        Returns:
            The output of the Encoder (batch_size, length, model_size)
            The alignment (attention) vectors for all layers
        """
        embed_out = self.embedding(sequence)

        embed_out *= tf.math.sqrt(tf.cast(self.model_size, tf.float32))
        embed_out += self.pes[:sequence.shape[1], :]
        embed_out = self.embedding_dropout(embed_out)

        sub_in = embed_out
        alignments = []

        for i in range(self.num_layers):
            sub_out, alignment = self.attention[i](sub_in, sub_in, encoder_mask)
            sub_out = self.attention_dropout[i](sub_out, training=training)
            sub_out = sub_in + sub_out
            sub_out = self.attention_norm[i](sub_out)
            
            alignments.append(alignment)
            ffn_in = sub_out

            ffn_out = self.dense_2[i](self.dense_1[i](ffn_in))
            ffn_out = self.ffn_dropout[i](ffn_out, training=training)
            ffn_out = ffn_in + ffn_out
            ffn_out = self.ffn_norm[i](ffn_out)

            sub_in = ffn_out

        return ffn_out, alignments


#### decoder ###################################################################

class Decoder(tf.keras.Model):
    """ Class for the Decoder
    Args:
        vocab_size: vocabulary size of the output
        model_size: d_model in the paper (depth size of the model)
        num_layers: number of layers (Multi-Head Attention + FNN)
        h: number of attention heads
        pes: positional encodings

        embedding: Embedding layer
        embedding_dropout: Dropout layer for Embedding
        attention_bot: array of bottom Multi-Head Attention layers (self attention)
        attention_bot_dropout: array of Dropout layers for bottom Multi-Head Attention
        attention_bot_norm: array of LayerNorm layers for bottom Multi-Head Attention
        attention_mid: array of middle Multi-Head Attention layers
        attention_mid_dropout: array of Dropout layers for middle Multi-Head Attention
        attention_mid_norm: array of LayerNorm layers for middle Multi-Head Attention
        dense_1: array of first Dense layers for FFN
        dense_2: array of second Dense layers for FFN
        ffn_dropout: array of Dropout layers for FFN
        ffn_norm: array of LayerNorm layers for FFN
        dense: Dense layer to compute final output
    """
    def __init__(self, vocab_size, model_size, num_layers, h, pes):
        super(Decoder, self).__init__()
        self.model_size = model_size
        self.num_layers = num_layers
        self.h = h
        self.pes = pes
        self.embedding = tf.keras.layers.Embedding(vocab_size, model_size)
        self.embedding_dropout = tf.keras.layers.Dropout(0.1)
        self.attention_bot = [MultiHeadAttention(model_size, h) for _ in range(num_layers)]
        self.attention_bot_dropout = [tf.keras.layers.Dropout(0.1) for _ in range(num_layers)]
        self.attention_bot_norm = [tf.keras.layers.LayerNormalization(
            epsilon=1e-6) for _ in range(num_layers)]
        self.attention_mid = [MultiHeadAttention(model_size, h) for _ in range(num_layers)]
        self.attention_mid_dropout = [tf.keras.layers.Dropout(0.1) for _ in range(num_layers)]
        self.attention_mid_norm = [tf.keras.layers.LayerNormalization(
            epsilon=1e-6) for _ in range(num_layers)]

        self.dense_1 = [tf.keras.layers.Dense(
            model_size * 4, activation='relu') for _ in range(num_layers)]
        self.dense_2 = [tf.keras.layers.Dense(
            model_size) for _ in range(num_layers)]
        self.ffn_dropout = [tf.keras.layers.Dropout(0.1) for _ in range(num_layers)]
        self.ffn_norm = [tf.keras.layers.LayerNormalization(
            epsilon=1e-6) for _ in range(num_layers)]

        self.dense = tf.keras.layers.Dense(vocab_size)

    def call(self, sequence, encoder_output, training=True, encoder_mask=None):
        """ Forward pass for the Decoder
        Args:
            sequence: source input sequences
            encoder_output: output of the Encoder (for computing middle attention)
            training: whether training or not (for Dropout)
            encoder_mask: padding mask for the Encoder's Multi-Head Attention
        
        Returns:
            The output of the Encoder (batch_size, length, model_size)
            The bottom alignment (attention) vectors for all layers
            The middle alignment (attention) vectors for all layers
        """
        # EMBEDDING AND POSITIONAL EMBEDDING
        embed_out = self.embedding(sequence)

        embed_out *= tf.math.sqrt(tf.cast(self.model_size, tf.float32))
        embed_out += self.pes[:sequence.shape[1], :]
        embed_out = self.embedding_dropout(embed_out)

        bot_sub_in = embed_out
        bot_alignments = []
        mid_alignments = []

        for i in range(self.num_layers):
            # BOTTOM MULTIHEAD SUB LAYER
            seq_len = bot_sub_in.shape[1]

            if training:
                mask = tf.linalg.band_part(tf.ones((seq_len, seq_len)), -1, 0)
            else:
                mask = None
            bot_sub_out, bot_alignment = self.attention_bot[i](bot_sub_in, bot_sub_in, mask)
            bot_sub_out = self.attention_bot_dropout[i](bot_sub_out, training=training)
            bot_sub_out = bot_sub_in + bot_sub_out
            bot_sub_out = self.attention_bot_norm[i](bot_sub_out)
            
            bot_alignments.append(bot_alignment)

            # MIDDLE MULTIHEAD SUB LAYER
            mid_sub_in = bot_sub_out

            mid_sub_out, mid_alignment = self.attention_mid[i](
                mid_sub_in, encoder_output, encoder_mask)
            mid_sub_out = self.attention_mid_dropout[i](mid_sub_out, training=training)
            mid_sub_out = mid_sub_out + mid_sub_in
            mid_sub_out = self.attention_mid_norm[i](mid_sub_out)
            
            mid_alignments.append(mid_alignment)

            # FFN
            ffn_in = mid_sub_out

            ffn_out = self.dense_2[i](self.dense_1[i](ffn_in))
            ffn_out = self.ffn_dropout[i](ffn_out, training=training)
            ffn_out = ffn_out + ffn_in
            ffn_out = self.ffn_norm[i](ffn_out)

            bot_sub_in = ffn_out

        logits = self.dense(ffn_out)

        return logits, bot_alignments, mid_alignments


#### loss function #############################################################

crossentropy = tf.keras.losses.SparseCategoricalCrossentropy(
    from_logits=True)


def loss_func(targets, logits):
    mask = tf.math.logical_not(tf.math.equal(targets, 0))
    mask = tf.cast(mask, dtype=tf.int64)
    loss = crossentropy(targets, logits, sample_weight=mask)

    return loss


class WarmupThenDecaySchedule(tf.keras.optimizers.schedules.LearningRateSchedule):
    """ Learning schedule for training the Transformer
    Attributes:
        model_size: d_model in the paper (depth size of the model)
        warmup_steps: number of warmup steps at the beginning
    """
    def __init__(self, model_size, warmup_steps=4000):
        super(WarmupThenDecaySchedule, self).__init__()

        self.model_size = model_size
        self.model_size = tf.cast(self.model_size, tf.float32)

        self.warmup_steps = warmup_steps

    def __call__(self, step):
        step_term = tf.math.rsqrt(step)
        warmup_term = step * (self.warmup_steps ** -1.5)

        return tf.math.rsqrt(self.model_size) * tf.math.minimum(step_term, warmup_term)


lr = WarmupThenDecaySchedule(MODEL_SIZE)
optimizer = tf.keras.optimizers.Adam(lr,
                                     beta_1=0.9,
                                     beta_2=0.98,
                                     epsilon=1e-9)


#### train step ################################################################

@tf.function
def train_step(source_seq, target_seq_in, target_seq_out):
    """ Execute one training step (forward pass + backward pass)
    Args:
        source_seq: source sequences
        target_seq_in: input target sequences (<start> + ...)
        target_seq_out: output target sequences (... + <end>)
    
    Returns:
        The loss value of the current pass
    """
    with tf.GradientTape() as tape:
        encoder_mask = 1 - tf.cast(tf.equal(source_seq, 0), dtype=tf.float32)
        # encoder_mask has shape (batch_size, source_len)
        # we need to add two more dimensions in between
        # to make it broadcastable when computing attention heads
        encoder_mask = tf.expand_dims(encoder_mask, axis=1)
        encoder_mask = tf.expand_dims(encoder_mask, axis=1)
        encoder_output, _ = encoder(source_seq, encoder_mask=encoder_mask)

        decoder_output, _, _ = decoder(
            target_seq_in, encoder_output, encoder_mask=encoder_mask)

        loss = loss_func(target_seq_out, decoder_output)

    variables = encoder.trainable_variables + decoder.trainable_variables
    gradients = tape.gradient(loss, variables)
    optimizer.apply_gradients(zip(gradients, variables))

    return loss


#### predict ###################################################################

def predict(test_source_text=None):
    """ Predict the output sentence for a given input sentence
    Args:
        test_source_text: input sentence (raw string)
    Returns:
        The output string array
    """
    if test_source_text is None:
        test_source_text = raw_data_old[np.random.choice(len(raw_data_old))]
    test_source_seq = old_tokenizer.texts_to_sequences([test_source_text])

    en_output, en_alignments = encoder(
        tf.constant(test_source_seq), training=False)

    de_input = tf.constant(
        [[new_tokenizer.word_index['<start>']]], dtype=tf.int64)

    out_words = []

    while True:
        de_output, de_bot_alignments, de_mid_alignments =  decoder(
            de_input, en_output, training=False)
        new_word = tf.expand_dims(tf.argmax(de_output, -1)[:, -1], axis=1)
        out_words.append(new_tokenizer.index_word[new_word.numpy()[0][0]])

        # Transformer doesn't have sequential mechanism (i.e. states)
        # so we have to add the last predicted word to create a new input 
        # sequence
        de_input = tf.concat((de_input, new_word), axis=-1)

        # TODO: get a nicer constraint for the sequence length!
        if out_words[-1] == '<end>' or len(out_words) >= max_length:
            break

    return out_words


def test_predict():
    print("---- testing predictions ----")
    for fragment in raw_data_old:
        input = fragment
        print("in  :", input)
        words = predict(input)
        print("out :", ' '.join(words), "\n")


#### training loop #############################################################

H          = 8
NUM_LAYERS = 4
NUM_EPOCHS = 200

en_vocab_size = len(old_tokenizer.word_index) + 1
nl_vocab_size = len(new_tokenizer.word_index) + 1

encoder = Encoder(en_vocab_size, MODEL_SIZE, NUM_LAYERS, H, pes)
decoder = Decoder(nl_vocab_size, MODEL_SIZE, NUM_LAYERS, H, pes)

print('Input  (old) vocabulary size', en_vocab_size)
print('Output (new) vocabulary size', nl_vocab_size)

start_time = time.time()
for e in range(NUM_EPOCHS):
    count = 0
    avg_loss = 0.0
    for batch, (src_seq, tgt_seq_in, tgt_seq_out) in enumerate(dataset.take(-1)):
        """
        print("encoder_in:\n", src_seq.numpy())
        print("decoder_in:\n", tgt_seq_in.numpy())
        print("decoder_out:\n", tgt_seq_out.numpy())
`       """
        loss = train_step(src_seq, tgt_seq_in, tgt_seq_out)
        avg_loss += loss
        count += 1

    print("epoch %d loss %.3f\r"%(e+1, loss.numpy()), end="")
    sys.stdout.flush()
    avg_loss /= float(count)

    if (e + 1) % 10 == 0:
        end_time = time.time()
        print('epoch %d: average loss %.3f dt: %.2fs'%(
            e+1, avg_loss, (end_time-start_time)/(e+1)))
        try:
            predict()
        except Exception as e:
            print(e)
            continue

test_predict()

--------------------------------------------------------------------------------

I hope this is of any use to you! Unfortunately the loss goes down equally fast with tf 2.1 as with tf 2.2, but you will see that the final results are very different...

**Comment by lordtt13** [comment]
I tried both scripts out, [TF 2.2.0](https://gist.github.com/lordtt13/0a96255e78e5009bde394b3e588801e1) and [TF 2.1.0](https://gist.github.com/lordtt13/f94e4fba62d311e8483f17b4e0cd3a9e). A rather obvious suggestion would be to try out a bigger dataset because overfitting may have been causing these issues.

**Comment by notnot** [comment]
at this point in time, it is exactly the overfitting that i want to see
as a sort of verification that the model is doing what i expect it to do

**Comment by notnot** [comment]
Ah yes, now i see your results... they are very much like mine!
Thanks for confirming. Overfitting or not, i would expect two versions of
tensorflow to give similar results.

**Comment by lordtt13** [comment]
Yeah, I'm stumped too.

**Comment by Saduf2019** [comment]
@notnot 
I ran the code on 2.2 andd 2.1 please find the [gist for 2.2](https://colab.research.google.com/gist/Saduf2019/7629336bee8b00a87187bfc17fed04e4/untitled307.ipynb) and [2.1](https://colab.research.google.com/gist/Saduf2019/277384e19026ed2d13e571de241a6cc7/untitled308.ipynb) please let us know if it confirms the issuereported.

**Comment by notnot** [comment]
Yes this totally confirms it. Worst results i have ever seen, with tf 2.2,
while the tf 2.1 results are perfect...

**Comment by lordtt13** [comment]
@notnot please check with tf2.3, the issue has been fixed now, check [here](https://gist.github.com/lordtt13/c8e6870408c656dc40bc2d9b884ba0d6)

**Comment by ymodak** [comment]
Thanks for your issue. This is fixed TF 2.3 now.
See the [gist](https://colab.research.google.com/gist/ymodak/e77a894202fe7c7e4d769b7d8deab0ea/untitled307.ipynb)
